[WIP]proposal: Occupied capacity increment rate (#71)
nervosnetwork/rfcs: rfcs/0000-occupied-capacity-increment-rate/0000-occupied-capacity-increment-rate.md

---
Number: 0000
Category: Standards Track
Status: Proposal
Author: Jinyang Jiang
Organization: Nervos Foundation
Created: 2019-2-11
---

# Occupied capacity increment rate

## Abstract

Capacity is one of the core resources in CKB network, itself is just a number representing how much storage resource a user can use.

Once a user decides to store something in CKB, the user must lock some capacity in exchange, the locked capacity which we called occupied capacity represents the actually used resource in the network.

Occupied capacity affects CKB network in two ways:

1. Single block occupied capacity: if a single block or transaction contains too much occupied capacity, it may affect the propagation of the block and may slow down the whole network.
2. The whole network occupied capacity: that every full node must cost the same memory or disk to save these content, it incentives miners to be more centralized which does not make any sense for a decentralized network.

> So the solution is to limit occupied capacity in each block, to indirectly control the network's total occupied capacity and the increment rate of occupied capacity.

Reviewer: Should be 'limit the increase of occupied capacity'.

This RFC proposes to make occupied capacity under control by adding two verification rules:

1. limit the block size

> 2. limit the increment of total network occupied capacity in block level

Reviewer: These are separate concerns, should be discussed in different RFCs.

## Specification

Define three variables:

* `MAX_BLOCK_BODY_BYTES`           - the max bytes of `commit_transactions` and `proposal_transactions` in a block, value: 10 MB

> * `MAX_BLOCK_UNCLES_BYTES`         - the max bytes of uncles blocks, value: 1 MB

Reviewer: Why do we need an additional `MAX_BLOCK_UNCLES_BYTES`? What's the problem with using only `MAX_BLOCK_BODY_BYTES`?

* `MAX_INCREMENT_OCCUPIED_CAPACITY` - the max increment occupied capacity of a block, value: 2 MB

Add two verification rules to CKB consensus:

### 1. Block size verification rule

Unfortunately, we could not just put a block size limitation on CKB like Bitcoin.

Because CKB has four fields defined in each block: `header`, `uncles`, `commit_transactions`, `proposal_transactions`, but CKB miner takes fees only from `commit_transactions`, they do not receive any incentives for submitting other parts. So a reasonable miner first packs `commit_transactions` into the block to make sure transactions fees as more as possible and immediately give up packs other parts once block size reaches the limitation.

It is bad news because CKB consensus adjusts the network based on uncle blocks, CKB consensus will not properly work once miners give up to submit `uncles`.

> So instead of a simple block size limitation, we must separately limit the size of `uncles` from other parts to make CKB consensus work under the designed intention.

Reviewer: Or incentivize uncles inclusion. This RFC should not be merged before we settled consensus details.

Author: I'll remove the "limit block size" part from the proposal and make another PR only include part of the "limit increase of occupied capacity".

---
Number: 0000
Category: Standards Track
Status: Proposal
Author: Jinyang Jiang
Organization: Nervos Foundation
Created: 2019-2-11
---

# Occupied capacity increment rate

## Abstract

Capacity is one of the core resources in CKB network, itself is just a number representing how much storage resource a user can use.

> Once a user decides to store something in CKB, the user must lock some capacity in exchange, the locked capacity which we called occupied capacity represent the actually used resource in the network.

Reviewer 2: `represent` -> `represents`

Occupied capacity affects CKB network in two ways:

1. Single block occupied capacity: if a single block or transaction contains too much occupied capacity, it may affect the propagation of the block and may slow down the whole network.

> 2. The whole network occupied capacity: that every full node must cost the same memory or disk to save these content, it incentive miners more centralize which is not making any sense for a decentralized network.

Reviewer 2: ```suggestion
2. The whole network occupied capacity: that every full node must cost the same memory or disk to save these content. It incentives miners to be more centralized, which does not make any sense for a decentralized network.
```

So the solution is to limit occupied capacity in each block, to indirectly control the network's total occupied capacity and the increment rate of occupied capacity.

## Specification

Define two variables:

* `MAX_BLOCK_BYTES`                 - the max size of a block, value: 10 MB

> * `MAX_INCREMENT_OCCUPIED_CAPACITY` - the max increment occupied capacity of a block, value: 2 MB

Reviewer 2: This value must be enough to run the demo, cc @xxuejie

Add two verification rules to CKB consensus:

### 1. Block size verification rule

> Serialized block size must smaller or equal than `MAX_BLOCK_BYTES`.

Reviewer 2: ```suggestion
Serialized block size must be smaller than or equal to `MAX_BLOCK_BYTES`.
```

Pseudo code: 
``` rust
assert!(serialized_block_size <= MAX_BLOCK_BYTES)
```

> CKB use [CFB][1] as the serializer, consider the serializer formats should not affect consensus rule, the `serialized_block_size` should be the actual byte size of data that do not include metadata introduced by CFB.

Reviewer 2: ```suggestion
CKB uses [CFB][1] as the serializer. Considering that the serializer format should not affect consensus rule, the `serialized_block_size` should be the actual byte size of data excluding metadata introduced by CFB.
```

Reviewer 2: It is unclear what is the metadata introduced by CFB.

Author: Flatbuffer encodes extra information into serialized binary to describe the struct(for example length of a list). So I assume CFB will do the same thing.

Which I mean is block size consensus rule should not consider that extra information (metadata); instead, we should verify the actual size(remove padding and metadata which CFB introduced) of a block.

Author: Bitcoin uses serialized size in consensus rule, but in our case, that CFB is a memory efficient serialization library and it is more complicated than bitcoin's serialization.
It is mean it will be hard to upgrade CFB format if the consensus rule depends on CFB serialization.

> This rule makes a single block occupied capacity under control, and also solving several other problems (unlock script size is too big, arguments size is too long, ...).

Reviewer 2: ```suggestion
This rule makes a single block occupied capacity under control, and also solves several other problems (unlock script size is too big, arguments size is too long, ...).
```

### 2. Increment occupied capacity verification rule

Define `increment_occupied_capacity` as the sum of all outputs occupied capacity in the block minus all inputs occupied capacity in the block.

> The occupied capacity that a block introduced should less or equal than `MAX_INCREMENT_OCCUPIED_CAPACITY`.

Reviewer 2: ```suggestion
The occupied capacity that a block introduced must be less than or equal to `MAX_INCREMENT_OCCUPIED_CAPACITY`.
```

Pseudo code:
``` rust
let mut inputs_occupied_capacity = 0;
let mut outputs_occupied_capacity = 0;
for tx in block.commit_transactions() {

>   inputs_occupied_capacity += tx.inputs.map(|input| input.occupied_capacity()).sum();

Reviewer 2: Better explain that input occupied capacity really means the occupied capacity of the output it refers to.

  outputs_occupied_capacity += tx.outputs.map(|output| output.occupied_capacity()).sum();
}

assert!(outputs_occupied_capacity - inputs_occupied_capacity <= MAX_INCREMENT_OCCUPIED_CAPACITY);
```

### Conclusion

The two rules make the increment of the whole network occupied capacity under control, and leaves room for other RFCs to make proposals like adjusting these two values via consensus (DAO or soft-fork).

## References

1. [CFB][1]

> [1]: https://github.com/nervosnetwork/rfcs/pull/47

Reviewer 2: Use link to RFC0008
